### DISCUSSION TITLE
Are STIGID and CIS identifiers or references?

### DISCUSSION THREAD

**Discussion by carlosmmatos** [ComplianceAsCode/content#6920] [Ideas]
per issue #6905 - @cipherboy posted the following:

> If I understand correctly, references have historically been to external benchmarks. Identifiers have to be unique to a rule though.
> 
> Typically though, Identifiers are product-specific. Some products (such as SUSE and RHEL) have CCE numbers and assign them. Functionally, AFAICT, CCEs are the only identifiers in the build system.
> 
> References (with the exception of stigid and cis) are typically product-independent. They're not unique and multiple rules might have the same reference number and one rule might have multiple reference numbers.
> 
> But does this distinction matter?
> 
> Does it make sense to move stigid and cis to the identifier section and drop the strict uniqueness check?
> 
> Especially with the auto-profile tool based on product-independent reference controls -- we'd either need to add product-specific-ness into this tooling if we wanted to add e.g., CIS or STIG (essentially, in my view, recreating profiles) -- or we could agree that the definitions of "identifier" and "references" are a bit loose, given that the rule_id is the real identifier ;-)
> 
> My 2c.

We'd like to open up a discussion with the community surrounding the use of references/identifiers and how we can better understand their relevance in the overall project.

**Comment by alexhaydock** [comment]
While it's very helpful to have the references/identifiers in the compliance output for ease-of-auditing, as someone who deals with CIS a lot, their references/identifiers are... not super helpful. Not only do they differ between product but they also differ between version within the same product. A quick glance at the Summary Table Appendix in the RHEL 7 and RHEL 8 CIS benchmarks shows that while most of the rules have the same identifiers when they appear in both, it's not always the case. To pick a quick example, `CIS 3.3.2` for RHEL 8 deals with disabling SCTP, while `CIS 3.3.2` for RHEL 7 is about ICMP redirects.

As long as we can provide every rule from the benchmark to audit against, I'm not 100% sure it's that useful to be carrying out all the extra legwork to make sure all the references and identifiers are present. Honestly, I'd personally be more interested in the static `rule_id` myself if it was made more visible and was consistent across products and versions.

**Reply by cipherboy** [reply]
I'm not sure I fully follow :) 

CIS and STIG are not only _product_-version tied but also _benchmark_-version tied. So for instance, CIS for Ubuntu 18.04 will have completely different versions than CIS for Ubuntu 20.04 (and that's expected)... ...but so might CIS v1.0.0 for Ubuntu 18.04 and CIS v2.0.1 for Ubuntu 18.04. You've gotta carefully correlate what is in the benchmark content you're running with the actual CIS version (and hopefully the CaC/... profile you're running tells you that info!). Once that's done, any discrepancies are bugs and should be filed as such. ;-)

Great idea exposing `rule_id` though, I'll have to think more about where that should go. :-) 

CIS I think has more trouble than STIG because STIG's identifiers try to be semi-unique whereas CIS is just top->down section numbering :/ 

----

Also, since it was mentioned in the comment on the issue but not here: maybe both sections could/should be combined into one?

**Reply by alexhaydock** [reply]
I may have misunderstood the original ask of the thread, but I think broadly I was agreeing with your original point here:

> we could agree that the definitions of "identifier" and "references" are a bit loose, given that the rule_id is the real identifier ;-)

As someone who uses CIS heavily, what I was getting at was that apart from a pure checklist-driven exercise, I don't find the CIS identifiers helpful at all for the reasons we've identified (different between products, different between product versions, different between benchmark versions). When you're auditing a heterogenous fleet with many different products and versions, the inconsistency between the identifiers is a pain. They are, as you say, "just top->down section numbering".

I can only speak for myself, but personally I'd benefit much more from static `rule_id`s being exposed more visibly since I'd rather rely on those.

I haven't been following CaC development closely enough to answer the question of whether it makes sense to move STIGID and CIS to the identifier section, but I was just adding my two cents with regard to the usefulness (or not) of the large amount of work that might be involved in "adding product-specific-ness" or recreating the profiles.

**Reply by vojtapolasek** [reply]
> I'm not sure I fully follow :)
> 
> CIS and STIG are not only _product_-version tied but also _benchmark_-version tied. So for instance, CIS for Ubuntu 18.04 will have completely different versions than CIS for Ubuntu 20.04 (and that's expected)... ...but so might CIS v1.0.0 for Ubuntu 18.04 and CIS v2.0.1 for Ubuntu 18.04. You've gotta carefully correlate what is in the benchmark content you're running with the actual CIS version (and hopefully the CaC/... profile you're running tells you that info!).

Does it make sense to care about outdated benchmark versions? I mean, would we have profile for rhel7 CIS 2.x and rhel7 CIS 3.x? I always assumed that we won't have that.

**Reply by cipherboy** [reply]
> I may have misunderstood the original ask of the thread, but I think broadly I was agreeing with your original point here:
> 
> > we could agree that the definitions of "identifier" and "references" are a bit loose, given that the rule_id is the real identifier ;-)
> 
> As someone who uses CIS heavily, what I was getting at was that apart from a pure checklist-driven exercise, I don't find the CIS identifiers helpful at all for the reasons we've identified (different between products, different between product versions, different between benchmark versions). When you're auditing a heterogenous fleet with many different products and versions, the inconsistency between the identifiers is a pain. They are, as you say, "just top->down section numbering".
> 
> I can only speak for myself, but personally I'd benefit much more from static `rule_id`s being exposed more visibly since I'd rather rely on those.
> 
> I haven't been following CaC development closely enough to answer the question of whether it makes sense to move STIGID and CIS to the identifier section, but I was just adding my two cents with regard to the usefulness (or not) of the large amount of work that might be involved in "adding product-specific-ness" or recreating the profiles.

I'm no XSLT expert but if you're willing to dive in... 

I [think here](https://github.com/ComplianceAsCode/content/blob/master/shared/transforms/shared_shorthand2xccdf.xslt#L222-L234) is about where you'd start adding it. Good luck though, I can only review the results if you get something working, I've got very little knowledge of how to add it from scratch! :rofl: 

---

Also, intent of this was mostly from the [`rule.yml`](https://complianceascode.readthedocs.io/en/latest/manual/developer/06_contributing_with_content.html#rules) side of things, I wasn't intending to change the output :-) But these are good ideas!

**Reply by cipherboy** [reply]
Hey @vojtapolasek :)

> > I'm not sure I fully follow :)
> > CIS and STIG are not only _product_-version tied but also _benchmark_-version tied. So for instance, CIS for Ubuntu 18.04 will have completely different versions than CIS for Ubuntu 20.04 (and that's expected)... ...but so might CIS v1.0.0 for Ubuntu 18.04 and CIS v2.0.1 for Ubuntu 18.04. You've gotta carefully correlate what is in the benchmark content you're running with the actual CIS version (and hopefully the CaC/... profile you're running tells you that info!).
>
> Does it make sense to care about outdated benchmark versions? I mean, would we have profile for rhel7 CIS 2.x and rhel7 CIS 3.x? I always assumed that we won't have that.

(And purely speaking for myself here...)

It isn't that we necessarily _care_ about outdated benchmark versions more so than sometimes we're sometimes _slow_ to update to the new version when it comes out. :-) Exposing to the end-consumer of our content what version we intended the CIS/STIG content to be at -- so regardless of CaC version they know what they're getting -- is probably best. My opinion is that they can always upgrade/downgrade release versions if they want a different version of content. Just gotta accept the bugs that were fixed in later versions. :-)

My 2c.

**Reply by vojtapolasek** [reply]
> Hey @vojtapolasek :)
> 
> > > I'm not sure I fully follow :)
> > > CIS and STIG are not only _product_-version tied but also _benchmark_-version tied. So for instance, CIS for Ubuntu 18.04 will have completely different versions than CIS for Ubuntu 20.04 (and that's expected)... ...but so might CIS v1.0.0 for Ubuntu 18.04 and CIS v2.0.1 for Ubuntu 18.04. You've gotta carefully correlate what is in the benchmark content you're running with the actual CIS version (and hopefully the CaC/... profile you're running tells you that info!).
> > 
> > 
> > Does it make sense to care about outdated benchmark versions? I mean, would we have profile for rhel7 CIS 2.x and rhel7 CIS 3.x? I always assumed that we won't have that.
> 
> (And purely speaking for myself here...)
> 
> It isn't that we necessarily _care_ about outdated benchmark versions more so than sometimes we're sometimes _slow_ to update to the new version when it comes out. :-) Exposing to the end-consumer of our content what version we intended the CIS/STIG content to be at -- so regardless of CaC version they know what they're getting -- is probably best. My opinion is that they can always upgrade/downgrade release versions if they want a different version of content. Just gotta accept the bugs that were fixed in later versions. :-)
> 
> My 2c.

I was just not sure if you are suggeting differentiating among CIS references based on benchmark versions. Are you? like cis@rhel7@2.0.1?

**Reply by cipherboy** [reply]
> I was just not sure if you are suggeting differentiating among CIS references based on benchmark versions. Are you? like cis@rhel7@2.0.1?

Ohhhh no. I just meant in the profile :D 

e.g., [from the draft PR for Ubuntu 20.04](https://github.com/ComplianceAsCode/content/pull/6416):
> ```yaml
> description: |-
>  This baseline aligns to the Center for Internet Security
>  Ubuntu 20.04 LTS Benchmark, v1.0.0, released 07-21-2020.
> ```

That's where I'd record the version number.

I think maintaining historical references (`cis@rhel7@2.0.1`) would be _wayyyy_ more work than I'd want to do ;-)

**Reply by yuumasato** [reply]
> I think maintaining historical references (`cis@rhel7@2.0.1`) would be _wayyyy_ more work than I'd want to do ;-)

Agree, maintaining historical references would be too much work. 

@alexhaydock 
> I can only speak for myself, but personally I'd benefit much more from static `rule_id`s being exposed more visibly since I'd rather rely on those.

Can you expand on this?
I'm not sure where you are headed, cause I think the HTML reports list the rule ID.

**Reply by alexhaydock** [reply]
> Can you expand on this?
> I'm not sure where you are headed, cause I think the HTML reports list the rule ID.

I think what I was thinking was mostly that CIS references/identifiers have been a pain point for me before personally when auditing a diverse environment, since the rule identifiers aren't consistent between systems or versions of the same system. But inside CaC, that static rule ID exists for each rule which is a great way to unify the same rule across all the different CIS benchmarks.

It's not something I'm doing _yet_ but I think for my own audit & reporting purposes I was thinking I might switch away from using CIS identifiers and towards using the CaC rule IDs. On that basis, I figured it would be nice if they were "more visible" in the HTML reports. Ideally in the main table alongside the Severity / Result columns. But the rule IDs can be quite long so, when I think about it, there's probably not a clean way of doing this without destroying the well-formatted HTML tables that currently exist. And you're right to point out that the rule ID is already available in the output if you look for it specifically by clicking the rule name. I think that will probably serve me well enough for my needs currently.

**Reply by yuumasato** [reply]
Not sure if you meant HTML tables or report. Some policies have HTML tables for the references, did you mean that?

RHEL products generate these tables for CIS for example: `build/tables/table-rhel8-cisresfs.html`
Does that seem useful?
Looking at it now, I see it doesn't list the rule ID, so that is probably what you meant, :)

**Reply by matejak** [reply]
FYI, are also new-style tables at `build/tables/tables-<prodname>-all.html` that generate the table HTML based on a Jinja template using the tiny [gen_tables.py](https://github.com/ComplianceAsCode/content/blob/master/utils/gen_tables.py) script. That script could be extended with little effort to be able to provide reference tables to support significant workflows.

Another thing that you may find interesting is that we contemplate to introduce a simplified JSON output for the OpenSCAP scanner. That output would come as an addition to current SCAP-compliant outputs, and it would make it easy to take the results, and process them in a higher-level application. See e.g. https://github.com/OpenSCAP/openscap/issues/1555 and if you have concrete ideas what such output should consist of, open a new feature request issue in the project.
